sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/RequestRetryPolicy.java

```diff
@@ -253,19 +253,17 @@ private HttpResponse attemptSync(final HttpPipelineCallContext context, HttpPipe
              * We want to send the request with a given timeout, but we don't want to kickoff that timeout-bound
              * operation until after the retry backoff delay, so we call delaySubscription.
              */
-            HttpResponse response = next.clone().processSync();
+            Mono<HttpResponse> httpResponseMono = Mono.fromCallable(() -> next.clone().processSync());
 
             // Default try timeout is Integer.MAX_VALUE seconds, if it's that don't set a timeout as that's about 68 years
             // and would likely never complete.
             // TODO (alzimmer): Think about not adding this if it's over a certain length, like 1 year.
             if (this.requestRetryOptions.getTryTimeoutDuration().getSeconds() != Integer.MAX_VALUE) {
-                try {
-                    Thread.sleep(this.requestRetryOptions.getTryTimeoutDuration().toMillis());
-                } catch (InterruptedException ie) {
-                    throw LOGGER.logExceptionAsError(new RuntimeException(ie));
-                }
+                httpResponseMono = httpResponseMono.timeout(this.requestRetryOptions.getTryTimeoutDuration());
```

Comment:
While this fixes the issue, this isn't the behavior we want. The `*Sync` methods shouldn't use Reactor code at all.

I believe the best choice here would be following what `azure-data-tables` is doing in their synchronous clients when there is a timeout: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/implementation/TableUtils.java#L615

@ibrahimrabab @ibrandes FYI

Reply:
The code you linked uses Executors.newCachedThreadPool() which can create an unbounded number of threads. Is that intentional? I've noticed similar custom code in other places too.

Seems like there are already uses of reactor code in the synchronous clients as well. Eg - HttpResponse.getBody() returns Flux<ByteBuffer>. So I assumed it was fine to use that approach since it is consistent with the async approach. 

Are there plans to move away from reactor for the sync flow? Curious about the reasons behind this.

Reply:
> The code you linked uses Executors.newCachedThreadPool() which can create an unbounded number of threads. Is that intentional?

I'll follow-up on this and get back to you.

> Seems like there are already uses of reactor code in the synchronous clients as well. Eg - HttpResponse.getBody() returns Flux. So I assumed it was fine to use that approach since it is consistent with the async approach.
>
> Are there plans to move away from reactor for the sync flow? Curious about the reasons behind this.

`HttpResponse.getBody()` is an unfortunate case of legacy APIs and doing a softer landing migration on removing Reactor usage from synchronous code flows.

While Reactor provides a lot of great functionality, if you're not familiar with it troubleshooting exceptions can be difficult. Due to our sync-over-async pattern we used to use exception stack traces would be obfuscated with Reactor stack elements, which we received strong feedback on debuggability, leading to more issues filed which some were simple errors once you knew how to parse the stack trace. Additionally, Reactor does a lot of "magic" for you and from testing having a fully synchronous flow when making a synchronous call (or as best as possible as some of the HTTP stacks we support are async only) resulted in better performance.

So, between debugability, coding convenience, and performance we've as best as possible split async APIs to fully async (always have been this way) and sync APIs to fully sync.

Reply:
Another observation I had was that the QueueClient additionally supports specifying timeout per request (same implementation as TableUtils). So we would submit to two threadPools to handle timeouts, one in RequestRetryPolicy and then in QueueClient. Would be good to eventually override the client request timeout with the per request timeout in a single place.

Currently working at Microsoft and happy to help chat, get on a call and contribute to this.

Reply:
> Another observation I had was that the QueueClient additionally supports specifying timeout per request (same implementation as TableUtils). So we would submit to two threadPools to handle timeouts, one in RequestRetryPolicy and then in QueueClient. Would be good to eventually override the client request timeout with the per request timeout in a single place.
> 
> Currently working at Microsoft and happy to help chat, get on a call and contribute to this.

This is a fantastic idea, would you mind filing an issue for this?

Reply:
> Another observation I had was that the QueueClient additionally supports specifying timeout per request (same implementation as TableUtils). So we would submit to two threadPools to handle timeouts, one in RequestRetryPolicy and then in QueueClient. Would be good to eventually override the client request timeout with the per request timeout in a single place.
> 
> Currently working at Microsoft and happy to help chat, get on a call and contribute to this.

Yeah, this would unfortunately result in multiple thread switches but the client API level timeout is an overall timeout, including all retries. Whereas the policy timeout is for the specific request. Both allow for a scenario like:

- I want to make sure this API call can only run for 10 seconds but I don't want to see any request take longer than 2 seconds.

Reply:
> I'll follow-up on this and get back to you.
> 
> >

Let me know the outcome of that. There are two similar implementations in the StorageImplUtils/TableUtils classes which use that approach. Happy to make changes if needed based on that decision.